timemachine/cpp/src/kernels/chiral_utils.cuh

```diff
@@ -0,0 +1,189 @@
+
+template <typename RealType> __device__ struct Matrix { RealType x0, y0, z0, x1, y1, z1, x2, y2, z2; };
+
+template <typename RealType> __device__ struct Vector {
+
+    RealType x, y, z;
+
+    __device__ RealType dot(Vector<RealType> const &other) { return x * other.x + y * other.y + z * other.z; }
+
+    __device__ RealType norm() const { return std::sqrt(x * x + y * y + z * z); }
+
+    __device__ Vector<RealType> unit() const {
+        RealType n = norm();
+        return {x / n, y / n, z / n};
+    }
+
+    __device__ Matrix<RealType> unit_jac() const {
+        RealType n = this->norm();
+        Vector<RealType> v = this->unit();
+        return {
+            (1 - v.x * v.x) / n,
+            (-v.x * v.y) / n,
+            (-v.x * v.z) / n,
+            (-v.y * v.x) / n,
+            (1 - v.y * v.y) / n,
+            (-v.y * v.z) / n,
+            (-v.z * v.x) / n,
+            (-v.z * v.y) / n,
+            (1 - v.z * v.z) / n};
+    }
+
+    __device__ Vector<RealType> operator+(Vector<RealType> const &obj) const {
+        return {x + obj.x, y + obj.y, z + obj.z};
+    }
+
+    __device__ Vector<RealType> operator-(Vector<RealType> const &obj) const {
+        return {x - obj.x, y - obj.y, z - obj.z};
+    }
+
+    // unary minus
+    __device__ Vector<RealType> operator-() const { return {-x, -y, -z}; }
+};
+
+template <typename RealType> __device__ Vector<RealType> mul_vm(Vector<RealType> v, Matrix<RealType> m) {
+
+    RealType v0 = v.x * m.x0 + v.y * m.x1 + v.z * m.x2;
+    RealType v1 = v.x * m.y0 + v.y * m.y1 + v.z * m.y2;
+    RealType v2 = v.x * m.z0 + v.y * m.z1 + v.z * m.z2;
+
+    return Vector<RealType>({v0, v1, v2});
+}
+
+template <typename RealType> __device__ Vector<RealType> cross_product(Vector<RealType> a, Vector<RealType> b) {
+
+    RealType cx = a.y * b.z - a.z * b.y;
+    RealType cy = a.z * b.x - a.x * b.z;
+    RealType cz = a.x * b.y - a.y * b.x;
+
+    return Vector<RealType>({cx, cy, cz});
+}
+
+template <typename RealType> __device__ Vector<RealType> difference(Vector<RealType> a, Vector<RealType> b) {
```

Comment:
Is this redundant with `operator-` defined above? (also seems unused)

Reply:
good call, will fix